dynamic/internal/shim/loader.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shim
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/apparentlymart/go-versions/versions"
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-mux/tf5to6server"
+	disco "github.com/hashicorp/terraform-svchost/disco"
+	"github.com/opentofu/opentofu/internal/getproviders"
+	"github.com/opentofu/opentofu/internal/logging"
+	tfplugin "github.com/opentofu/opentofu/internal/plugin"
+	"github.com/opentofu/opentofu/internal/providercache"
+	"github.com/opentofu/opentofu/internal/tfplugin5"
+	"github.com/opentofu/opentofu/internal/tfplugin6"
+	v5shim "github.com/opentofu/opentofu/shim/protov5"
+	v6shim "github.com/opentofu/opentofu/shim/protov6"
+	tfaddr "github.com/opentofu/registry-address"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// TODO:
+//
+// - If TF is setup and has already cached a provider, we should try to use that
+// provider.
+//
+// - If not, but TF is setup we should probably contribute to the existing cache.
+//
+// - If TF is not setup, we should cache within PULUMI_HOME to avoid creating new dirs.
+const envPluginCache = "TF_PLUGIN_CACHE_DIR"
```

Comment:
Should this be prefixed with PULUMI?

Reply:
It should be prefixed with Pulumi. Thanks for pointing that out.